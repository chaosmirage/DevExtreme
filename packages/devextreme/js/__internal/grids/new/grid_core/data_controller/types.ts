export type DataObject = Record<string, unknown>;
export type Key = unknown;
export type KeyExpr = unknown;
