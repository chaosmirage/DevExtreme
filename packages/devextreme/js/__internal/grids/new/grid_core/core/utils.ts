import { domAdapter } from '@ts/core/m_dom_adapter';
import type { InfernoNode } from 'inferno';
import { render } from 'inferno';

export function renderToDOMNode(input: InfernoNode): HTMLElement {
  const container = domAdapter.createElement('div');
  render(input, container);

  // eslint-disable-next-line max-len
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-explicit-any
  return container.children[0] as any;
}
