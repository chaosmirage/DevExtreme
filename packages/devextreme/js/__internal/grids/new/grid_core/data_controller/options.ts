import type { DataSourceLike } from '@js/data/data_source';

interface PagingOptions {
  enabled?: boolean;
  pageSize?: number;
  pageIndex?: number;
}

interface RemoteOperationsOptions {
  filtering?: boolean;
  paging?: boolean;
  sorting?: boolean;
  summary?: boolean;
}

export interface Options {
  cacheEnabled?: boolean;
  dataSource?: DataSourceLike<unknown>;
  dateSerializationFormat?: string;
  keyExpr?: string | string[];
  onDataErrorOccurred?: unknown;
  paging?: PagingOptions;
  remoteOperations?: RemoteOperationsOptions | boolean;
}

export const defaultOptions = {
  paging: {
    enabled: true,
    pageSize: 6,
    pageIndex: 0,
  },
  remoteOperations: {
    filtering: false,
    paging: false,
    sorting: false,
    summary: false,
  },
} satisfies Options;
