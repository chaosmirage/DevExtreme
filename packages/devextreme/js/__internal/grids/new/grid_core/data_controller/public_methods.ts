/* eslint-disable consistent-return */
/* eslint-disable @typescript-eslint/no-invalid-void-type */
/* eslint-disable spellcheck/spell-checker */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import type { FilterDescriptor } from '@js/data';
import type DataSource from '@js/data/data_source';
import { keysEqual } from '@ts/data/m_utils';

import type { Constructor } from '../types';
import type { GridCoreNewBase } from '../widget';
import type { DataObject, Key } from './types';

export function PublicMethods<T extends Constructor<GridCoreNewBase>>(GridCore: T) {
  return class GridCoreWithDataController extends GridCore {
    public getDataSource(): DataSource {
      return this.dataController.dataSource.unreactive_get();
    }

    public byKey(key: Key): Promise<DataObject> | undefined {
      const items = this.getDataSource().items();
      const store = this.getDataSource().store();
      const keyExpr = store.key();

      const foundItem = items.find(
        (item) => keysEqual(keyExpr, key, this.keyOf(item)),
      );

      if (foundItem) {
        return Promise.resolve(foundItem);
      }

      return store.byKey(key);
    }

    public getFilter(): FilterDescriptor | FilterDescriptor[] {
      return this.getDataSource().filter();
    }

    public keyOf(obj: DataObject) {
      return this.dataController.getDataKey(obj);
    }

    public pageCount(): number {
      return this.dataController.pageCount.unreactive_get();
    }

    public pageSize(): number;
    public pageSize(value: number): void;
    public pageSize(value?: number): number | void {
      if (value === undefined) {
        return this.dataController.pageSize.unreactive_get();
      }
      this.dataController.pageSize.update(value);
    }

    public pageIndex(): number;
    public pageIndex(newIndex: number): void;
    public pageIndex(newIndex?: number): number | void {
      if (newIndex === undefined) {
        return this.dataController.pageIndex.unreactive_get();
      }
      // TODO: Promise<void> (jQuery or native)
      return this.dataController.pageIndex.update(newIndex);
    }

    public totalCount(): number {
      return this.dataController.totalCount.unreactive_get();
    }
  };
}
