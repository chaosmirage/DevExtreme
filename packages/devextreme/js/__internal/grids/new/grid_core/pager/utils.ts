import type { PagerVisible, PageSize, PageSizes } from './options';

// TODO: Need to fix case with runtime changes the allowedPageSizes property to 'auto'
export function calculatePageSizes(
  allowedPageSizes: PageSize[] | undefined,
  pageSizesConfig: PageSizes,
  pageSize: number,
): PageSizes {
  if (Array.isArray(pageSizesConfig)) {
    return pageSizesConfig;
  }
  if (Array.isArray(allowedPageSizes) && allowedPageSizes.includes(pageSize)) {
    return allowedPageSizes;
  }
  if (pageSizesConfig && pageSize > 1) {
    return [Math.floor(pageSize / 2), pageSize, pageSize * 2];
  }

  return [];
}

export function isVisible(
  visibleConfig: PagerVisible,
  pageCount: number,
): boolean {
  if (visibleConfig === 'auto') {
    return pageCount > 1;
  }
  return visibleConfig;
}
