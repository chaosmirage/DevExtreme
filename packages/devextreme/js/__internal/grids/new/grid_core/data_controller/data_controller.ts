/* eslint-disable @typescript-eslint/no-invalid-void-type */
/* eslint-disable no-param-reassign */
/* eslint-disable spellcheck/spell-checker */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { DataSource } from '@js/common/data';
import ArrayStore from '@js/common/data/array_store';
import type { SubsGets } from '@ts/core/reactive/index';
import {
  computed, effect, state,
} from '@ts/core/reactive/index';
import { createPromise } from '@ts/core/utils/promise';

// import { EditingController } from '../editing/controller';
// import type { Change } from '../editing/types';
import { OptionsController } from '../options_controller/options_controller';
import type { DataObject, Key } from './types';
import {
  getLocalLoadOptions,
  getStoreLoadOptions,
  isCustomStore,
  isLocalStore,
  normalizeDataSource,
  normalizeLocalOptions,
  normalizeRemoteOptions,
  updateItemsImmutable,
} from './utils';

export class DataController {
  private readonly pendingLocalOperations = {};

  private readonly loadedPromise = createPromise<void>();

  private readonly dataSourceConfiguration = this.options.oneWay('dataSource');

  private readonly keyExpr = this.options.oneWay('keyExpr');

  public readonly dataSource = computed(
    (dataSourceLike, keyExpr) => normalizeDataSource(dataSourceLike, keyExpr),
    [this.dataSourceConfiguration, this.keyExpr],
  );

  // TODO
  private readonly cacheEnabled = this.options.oneWay('cacheEnabled');

  private readonly pagingEnabled = this.options.twoWay('paging.enabled');

  public readonly pageIndex = this.options.twoWay('paging.pageIndex');

  public readonly pageSize = this.options.twoWay('paging.pageSize');

  private readonly remoteOperations = this.options.oneWay('remoteOperations');

  private readonly onDataErrorOccurred = this.options.action('onDataErrorOccurred');

  private readonly _items = state<DataObject[]>([]);

  public readonly items: SubsGets<DataObject[]> = this._items;

  private readonly _totalCount = state(0);

  public readonly totalCount: SubsGets<number> = this._totalCount;

  public readonly isLoading = state(false);

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public readonly filter = this.options.twoWay('filterValue');

  // public itemsWithChanges = computed(
  //   (items, changes: Change[] | undefined) => items.map((item) => (changes ?? []).filter(
  //     (change) => change.key === this.getDataKey(item),
  //   ).reduce((p, v) => ({
  //     // @ts-expect-error
  //     ...item, ...v.data,
  //   }), item)),
  //   [this.items, this.editing.changes],
  // );

  public readonly pageCount = computed(
    (totalCount, pageSize) => Math.ceil(totalCount / pageSize),
    [this.totalCount, this.pageSize],
  );

  private readonly normalizedRemoteOptions = computed(
    (remoteOperations, dataSource) => {
      const store = dataSource.store();
      return normalizeRemoteOptions(remoteOperations, isLocalStore(store), isCustomStore(store));
    },
    [this.remoteOperations, this.dataSource],
  );

  private readonly normalizedLocalOperations = computed(
    (normalizedRemoteOperations) => normalizeLocalOptions(normalizedRemoteOperations),
    [this.normalizedRemoteOptions],
  );

  public static dependencies = [OptionsController] as const;

  constructor(
    private readonly options: OptionsController,
  ) {
    effect(
      (dataSource) => {
        const changedCallback = (e?): void => {
          this.onChanged(dataSource, e);
        };
        const loadingChangedCallback = (): void => {
          this.isLoading.update(dataSource.isLoading());
        };
        const loadErrorCallback = (error: string): void => {
          const callback = this.onDataErrorOccurred.unreactive_get();
          callback({ error });
          changedCallback();
        };
        const customizeStoreLoadOptionsCallback = (e): void => {
          const localOptions = this.normalizedLocalOperations.unreactive_get();
          this.pendingLocalOperations[e.operationId] = getLocalLoadOptions(
            e.storeLoadOptions,
            localOptions,
          );
          e.storeLoadOptions = getStoreLoadOptions(
            e.storeLoadOptions,
            localOptions,
          );
        };

        const dataLoadedCallback = (e): void => {
          /*
            We use Deffered here because the code below is synchronous.
            customizeLoadResult callback does not support async code.
          */
          new ArrayStore(e.data).load(this.pendingLocalOperations[e.operationId]).done((data) => {
            e.data = data;
          }).fail((error) => {
            // @ts-expect-error
            e.data = new Deferred().reject(error);
          });
          this.pendingLocalOperations[e.operationId] = undefined;
        };

        if (dataSource.isLoaded()) {
          changedCallback();
        }
        dataSource.on('changed', changedCallback);
        dataSource.on('loadingChanged', loadingChangedCallback);
        dataSource.on('loadError', loadErrorCallback);

        // @ts-expect-error
        dataSource.on('customizeStoreLoadOptions', customizeStoreLoadOptionsCallback);
        // @ts-expect-error
        dataSource.on('customizeLoadResult', dataLoadedCallback);

        return (): void => {
          dataSource.off('changed', changedCallback);
          dataSource.off('loadingChanged', loadingChangedCallback);
          dataSource.off('loadError', loadErrorCallback);

          // @ts-expect-error
          dataSource.off('customizeStoreLoadOptions', customizeStoreLoadOptionsCallback);
          // @ts-expect-error
          dataSource.off('customizeLoadResult', dataLoadedCallback);
        };
      },
      [this.dataSource],
    );

    effect(
      () => {
        if (this.dataSource.unreactive_get().isLoaded()) {
          // eslint-disable-next-line @typescript-eslint/no-floating-promises
          this.dataSource.unreactive_get().load();
        }
      },
      [this.normalizedRemoteOptions],
    );

    effect(
      (dataSource, pageIndex, pageSize, filter, pagingEnabled) => {
        let someParamChanged = false;
        if (dataSource.pageIndex() !== pageIndex) {
          dataSource.pageIndex(pageIndex);
          someParamChanged ||= true;
        }
        if (dataSource.pageSize() !== pageSize) {
          dataSource.pageSize(pageSize);
          someParamChanged ||= true;
        }
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-boolean-literal-compare
        if (dataSource.requireTotalCount() !== true) {
          dataSource.requireTotalCount(true);
          someParamChanged ||= true;
        }
        if (dataSource.filter() !== filter) {
          dataSource.filter(filter);
          someParamChanged ||= true;
        }
        if (dataSource.paginate() !== pagingEnabled) {
          dataSource.paginate(pagingEnabled);
          someParamChanged ||= true;
        }

        if (someParamChanged || !dataSource.isLoaded()) {
          // eslint-disable-next-line @typescript-eslint/no-floating-promises
          dataSource.load();
        }
      },
      [this.dataSource, this.pageIndex, this.pageSize, this.filter, this.pagingEnabled],
    );
  }

  private onChanged(dataSource: DataSource, e): void {
    let items = dataSource.items() as DataObject[];

    if (e?.changes) {
      items = this._items.unreactive_get();
      items = updateItemsImmutable(items, e.changes, dataSource.store());
    }

    this._items.update(items);
    this.pageIndex.update(dataSource.pageIndex());
    this.pageSize.update(dataSource.pageSize());
    this._totalCount.update(dataSource.totalCount());
    this.loadedPromise.resolve();
  }

  public getDataKey(data: DataObject): Key {
    return this.dataSource.unreactive_get().store().keyOf(data);
  }

  public waitLoaded(): Promise<void> {
    return this.loadedPromise.promise;
  }
}
