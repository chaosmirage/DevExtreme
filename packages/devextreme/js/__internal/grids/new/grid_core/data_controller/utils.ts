/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { ArrayStore, CustomStore } from '@js/common/data';
import type { Store } from '@js/data';
import type { DataSourceLike } from '@js/data/data_source';
import DataSource from '@js/data/data_source';
import { normalizeDataSourceOptions } from '@js/data/data_source/utils';
import { applyBatch } from '@ts/data/m_array_utils';

import type { DataObject, OperationOptions } from './types';

export function normalizeDataSource(
  dataSourceLike: DataSourceLike<unknown, unknown> | null | undefined,
  keyExpr: string | string[] | undefined,
): DataSource<unknown, unknown> {
  if (dataSourceLike instanceof DataSource) {
    return dataSourceLike;
  }

  if (Array.isArray(dataSourceLike)) {
    // eslint-disable-next-line no-param-reassign
    dataSourceLike = {
      store: {
        type: 'array',
        data: dataSourceLike,
        key: keyExpr,
      },
    };
  }

  // TODO: research making second param not required
  return new DataSource(normalizeDataSourceOptions(dataSourceLike, undefined));
}

export function isLocalStore(store: Store): boolean {
  return store instanceof ArrayStore;
}

export function isCustomStore(store: Store): boolean {
  return store instanceof CustomStore;
}

export function normalizeRemoteOptions(
  remoteOperations: boolean | object | string,
  localStore: boolean,
  customStore: boolean,
): OperationOptions {
  if (remoteOperations instanceof String && remoteOperations !== 'auto') {
    throw new Error('Remote operations do not support any string values except \'auto\'');
  }

  const disabledAllRemoteOperations = {
    filtering: false,
    sorting: false,
    paging: false,
    summary: false,
  };
  if (remoteOperations === false) {
    return disabledAllRemoteOperations;
  }

  if (remoteOperations === 'auto') {
    if (localStore || customStore) {
      return disabledAllRemoteOperations;
    }
    return {
      filtering: true,
      sorting: true,
      paging: true,
      summary: false,
    };
  }

  const enabledAllRemoteOperations = {
    filtering: true,
    sorting: true,
    paging: true,
    summary: true,
  };
  if (remoteOperations === true) {
    return enabledAllRemoteOperations;
  }

  return remoteOperations as object;
}

export function normalizeLocalOptions(
  normalizedRemoteOperations: OperationOptions,
): OperationOptions {
  return {
    filtering: !normalizedRemoteOperations.filtering,
    sorting: !normalizedRemoteOperations.sorting,
    paging: !normalizedRemoteOperations.paging,
    summary: !normalizedRemoteOperations.summary,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export function getLocalLoadOptions(originOptions, localOperations) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const localLoadOptions = {} as any;
  if (localOperations.sorting) {
    localLoadOptions.sort = originOptions.sort;
  }
  if (localOperations.filtering) {
    localLoadOptions.filter = originOptions.filter;
  }
  if (localOperations.paging) {
    localLoadOptions.skip = originOptions.skip;
    localLoadOptions.take = originOptions.take;
  }
  if (localOperations.summary) {
    localLoadOptions.summary = originOptions.summary;
  }
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return localLoadOptions;
}

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export function getStoreLoadOptions(originOptions, localOperations) {
  const storeLoadOptions = originOptions;
  if (localOperations.sorting) {
    delete storeLoadOptions.sort;
  }
  if (localOperations.filtering) {
    delete storeLoadOptions.filter;
  }
  if (localOperations.paging) {
    delete storeLoadOptions.skip;
    delete storeLoadOptions.take;
  }
  if (localOperations.summary) {
    delete storeLoadOptions.summary;
  }
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return storeLoadOptions;
}

export function updateItemsImmutable(
  data: DataObject[],
  changes: any[],
  keyInfo: any,
): DataObject[] {
  // @ts-expect-error
  return applyBatch({
    keyInfo,
    data,
    changes,
    immutable: true,
  });
}
