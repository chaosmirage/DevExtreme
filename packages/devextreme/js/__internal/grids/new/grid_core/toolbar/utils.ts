import { extend } from '@js/core/utils/extend';
import { isDefined, isString } from '@js/core/utils/type';
import type { Item as BaseToolbarItem } from '@js/ui/toolbar';

import type { ToolbarItems } from './types';

export function isVisible(
  visibleConfig: boolean | undefined,
  items: ToolbarItems,
): boolean {
  if (visibleConfig === undefined) {
    return items.length > 0;
  }

  return visibleConfig;
}

interface ToolbarItem extends BaseToolbarItem {
  name?: string;
}

function normalizeToolbarItem(
  item: ToolbarItem | string,
  defaultButtonsMap: { [key: string]: ToolbarItem },
  defaultItemNames: readonly string[],
): ToolbarItem {
  let button = item;
  const defaultProps: ToolbarItem = { location: 'after' };

  if (isString(button)) {
    button = { name: button };
  }

  if (isDefined(button.name)) {
    if (isDefined(defaultButtonsMap[button.name])) {
      button = extend(true, {}, defaultButtonsMap[button.name], button);
    } else if (defaultItemNames.includes(button.name)) {
      button = { ...button, visible: false };
    }
  }

  return extend(true, {}, defaultProps, button) as ToolbarItem;
}

export function normalizeToolbarItems(
  defaultItems: (ToolbarItem & { name: string })[],
  userItems: (ToolbarItem | string)[] | undefined,
  defaultItemNames: readonly string[],
): ToolbarItem[] {
  if (!isDefined(userItems)) {
    return defaultItems;
  }

  const defaultButtonsMap = {};

  defaultItems.forEach((button) => {
    defaultButtonsMap[button.name] = button;
  });

  return userItems.map(
    (
      item: ToolbarItem | string,
    ) => normalizeToolbarItem(item, defaultButtonsMap, defaultItemNames),
  );
}
