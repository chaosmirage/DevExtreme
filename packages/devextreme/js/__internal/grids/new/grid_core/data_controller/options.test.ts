/* eslint-disable spellcheck/spell-checker */
import {
  afterAll,
  beforeAll,
  describe, expect, it, jest,
} from '@jest/globals';
import { CustomStore } from '@js/common/data';
import DataSource from '@js/data/data_source';
import { logger } from '@ts/core/utils/m_console';
import ArrayStore from '@ts/data/m_array_store';

import type { Options } from '../options';
import { OptionsControllerMock } from '../options_controller/options_controller.mock';
import { DataController } from './data_controller';

beforeAll(() => {
  jest.spyOn(logger, 'error').mockImplementation(() => {});
});
afterAll(() => {
  jest.restoreAllMocks();
});

const setup = (options: Options) => {
  const optionsController = new OptionsControllerMock(options);
  const dataController = new DataController(optionsController);

  return {
    optionsController,
    dataController,
  };
};

describe('Options', () => {
  describe('cacheEnabled', () => {
    const setupForCacheEnabled = ({ cacheEnabled }) => {
      const store = new ArrayStore({
        data: [
          { id: 1, value: 'value 1' },
          { id: 2, value: 'value 2' },
          { id: 3, value: 'value 3' },
        ],
        key: 'id',
      });

      jest.spyOn(store, 'load');

      const { dataController } = setup({
        cacheEnabled,
        dataSource: store,
        paging: {
          pageSize: 1,
        },
      });

      return { store, dataController };
    };

    describe('when it is false', () => {
      it('should skip caching requests', () => {
        const { store, dataController } = setupForCacheEnabled({
          cacheEnabled: false,
        });
        expect(store.load).toBeCalledTimes(1);

        dataController.pageIndex.update(1);
        expect(store.load).toBeCalledTimes(2);

        dataController.pageIndex.update(0);
        expect(store.load).toBeCalledTimes(3);
      });
    });

    describe('when it is true', () => {
      it.skip('should cache previously loaded pages', () => {});
      it.skip('should clear cache if not only pageIndex changed', () => {});
    });
  });

  describe('dataSourse', () => {
    describe('when it is dataSource instance', () => {
      it('should pass dataSource as is', () => {
        const dataSource = new DataSource({
          store: [{ a: 1 }, { b: 2 }],
        });

        const { dataController } = setup({ dataSource });

        expect(dataController.dataSource.unreactive_get()).toBe(dataSource);
      });
    });
    describe('when it is array', () => {
      it('should normalize to DataSource with given items', () => {
        const data = [{ a: 1 }, { b: 2 }];
        const { dataController } = setup({ dataSource: data });

        const dataSource = dataController.dataSource.unreactive_get();

        expect(dataSource).toBeInstanceOf(DataSource);
        expect(dataSource.items()).toEqual(data);
      });
    });
    describe('when it is empty', () => {
      it('should should normalize to empty DataSource', () => {
        const { dataController } = setup({});

        const dataSource = dataController.dataSource.unreactive_get();

        expect(dataSource).toBeInstanceOf(DataSource);
        expect(dataSource.items()).toHaveLength(0);
      });
    });
  });

  describe('keyExpr', () => {
    describe('when dataSource is array', () => {
      it('should be passed as key to DataSource', () => {
        const { dataController } = setup({
          dataSource: [{ myKeyExpr: 1 }, { myKeyExpr: 2 }],
          keyExpr: 'myKeyExpr',
        });

        const dataSource = dataController.dataSource.unreactive_get();
        expect(dataSource.key()).toBe('myKeyExpr');
      });
    });
    describe('when dataSource is DataSource instance', () => {
      it('should be ignored', () => {
        const { dataController } = setup({
          dataSource: new ArrayStore({
            key: 'storeKeyExpr',
            data: [{ storeKeyExpr: 1 }, { storeKeyExpr: 2 }],
          }),
          keyExpr: 'myKeyExpr',
        });

        const dataSource = dataController.dataSource.unreactive_get();
        expect(dataSource.key()).toBe('storeKeyExpr');
      });
    });
  });

  describe('onDataErrorOccurred', () => {
    it('should be called when load error happens', async () => {
      const onDataErrorOccurred = jest.fn();

      const { dataController } = setup({
        dataSource: new CustomStore({
          load() {
            return Promise.reject(new Error('my error'));
          },
        }),
        onDataErrorOccurred,
      });

      await dataController.waitLoaded();

      expect(onDataErrorOccurred).toBeCalledTimes(1);
      expect(onDataErrorOccurred.mock.calls[0]).toMatchSnapshot([{
        component: expect.any(Object),
      }]);
    });
  });

  describe('paging.enabled', () => {
    describe('when it is true', () => {
      it('should turn on pagination', () => {
        const { dataController } = setup({
          dataSource: [{ a: '1' }, { a: '2' }, { a: '3' }, { a: '4' }],
          paging: {
            enabled: true,
            pageSize: 2,
          },
        });

        const items = dataController.items.unreactive_get();
        expect(items).toHaveLength(2);
      });
    });
    describe('when it is false', () => {
      it('should turn on pagination', () => {
        const { dataController } = setup({
          dataSource: [{ a: '1' }, { a: '2' }, { a: '3' }, { a: '4' }],
          paging: {
            enabled: false,
            pageSize: 2,
          },
        });

        const items = dataController.items.unreactive_get();
        expect(items).toHaveLength(4);
      });
    });
  });

  describe('paging.pageIndex', () => {
    it('should change current page', () => {
      const { dataController, optionsController } = setup({
        dataSource: [{ a: '1' }, { a: '2' }, { a: '3' }, { a: '4' }],
        paging: {
          pageSize: 2,
          pageIndex: 1,
        },
      });

      let items = dataController.items.unreactive_get();
      expect(items).toEqual([{ a: '3' }, { a: '4' }]);

      optionsController.option('paging.pageIndex', 0);
      items = dataController.items.unreactive_get();
      expect(items).toEqual([{ a: '1' }, { a: '2' }]);
    });
  });

  describe('paging.pageSize', () => {
    it('should change size of current page', () => {
      const { dataController, optionsController } = setup({
        dataSource: [{ a: '1' }, { a: '2' }, { a: '3' }, { a: '4' }],
        paging: {
          pageSize: 2,
        },
      });

      let items = dataController.items.unreactive_get();
      expect(items).toEqual([{ a: '1' }, { a: '2' }]);

      optionsController.option('paging.pageSize', 3);
      items = dataController.items.unreactive_get();
      expect(items).toEqual([{ a: '1' }, { a: '2' }, { a: '3' }]);
    });
  });

  describe.skip('remoteOperations', () => {

  });
});
