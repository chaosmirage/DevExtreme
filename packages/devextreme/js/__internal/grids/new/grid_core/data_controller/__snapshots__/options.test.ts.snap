// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options onDataErrorOccurred should be called when load error happens 1`] = `
[
  {
    "component": Any<Object>,
    "error": [Error: my error],
  },
]
`;
